iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/CteScanOperator.java

```diff
@@ -0,0 +1,134 @@
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing,
+ *  * software distributed under the License is distributed on an
+ *  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  * KIND, either express or implied.  See the License for the
+ *  * specific language governing permissions and limitations
+ *  * under the License.
+ *
+ */
+
+package org.apache.iotdb.db.queryengine.execution.operator.source.relational;
+
+import org.apache.iotdb.commons.utils.TestOnly;
+import org.apache.iotdb.db.queryengine.execution.MemoryEstimationHelper;
+import org.apache.iotdb.db.queryengine.execution.operator.OperatorContext;
+import org.apache.iotdb.db.queryengine.execution.operator.source.SourceOperator;
+import org.apache.iotdb.db.queryengine.plan.planner.plan.node.PlanNodeId;
+import org.apache.iotdb.db.utils.cte.CteDataReader;
+import org.apache.iotdb.db.utils.cte.CteDataStore;
+import org.apache.iotdb.db.utils.cte.MemoryReader;
+
+import org.apache.tsfile.read.common.block.TsBlock;
+import org.apache.tsfile.utils.RamUsageEstimator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CteScanOperator implements SourceOperator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CteScanOperator.class);
+  private static final long INSTANCE_SIZE =
+      RamUsageEstimator.shallowSizeOfInstance(CteScanOperator.class);
+
+  private final OperatorContext operatorContext;
+  private final PlanNodeId sourceId;
+
+  private final CteDataStore dataStore;
+  private final CteDataReader dataReader;
+  private final int dataStoreRefCount;
+
+  public CteScanOperator(
+      OperatorContext operatorContext, PlanNodeId sourceId, CteDataStore dataStore) {
+    this.operatorContext = operatorContext;
+    this.sourceId = sourceId;
+    this.dataStore = dataStore;
+    this.dataReader = new MemoryReader(dataStore.getCachedData());
+    this.dataStoreRefCount = dataStore.increaseRefCount();
+  }
+
+  @Override
+  public TsBlock next() throws Exception {
+    if (dataReader == null) {
+      return null;
+    }
+    return dataReader.next();
+  }
+
+  @Override
+  public boolean hasNext() throws Exception {
+    if (dataReader == null) {
+      return false;
+    }
+    return dataReader.hasNext();
+  }
+
+  @Override
+  public void close() throws Exception {
+    try {
+      if (dataReader != null) {
+        dataReader.close();
+      }
+    } catch (Exception e) {
+      LOGGER.error("Fail to close fileChannel", e);
+    }
+  }
+
+  @Override
+  public boolean isFinished() throws Exception {
+    return !hasNextWithTimer();
+  }
+
+  @Override
+  public long calculateMaxPeekMemory() {
+    return calculateRetainedSizeAfterCallingNext() + calculateMaxReturnSize();
+  }
+
+  @Override
+  public long calculateMaxReturnSize() {
+    // The returned object is a reference to TsBlock in CteDataReader
+    return RamUsageEstimator.NUM_BYTES_OBJECT_REF;
```

Comment:
it should return the actual max return size among all the tsblocks in dataReader

Reply:
done